packages/core/src/controllers/ConvertController.ts

```diff
@@ -0,0 +1,664 @@
+      : '0'
+  },
+
+  async getInitialGasPrice() {
+    const res = await ConvertApiController.getGasPrice()
+    const instant = res.instant
+    const value = typeof instant === 'object' ? res.instant.maxFeePerGas : instant
+    const gasFee = BigInt(value)
+    const gasLimit = BigInt(INITIAL_GAS_LIMIT)
+    const gasPrice = this.calculateGasPriceInUSD(gasLimit, gasFee)
+    state.gasPriceInUSD = gasPrice
+  },
+
+  async refreshConvertValues() {
+    const { fromAddress, toTokenDecimals, toTokenAddress } = this.getParams()
+
+    if (fromAddress && toTokenAddress && toTokenDecimals && !state.loading) {
+      const transaction = await this.getTransaction()
+      this.setTransactionDetails(transaction)
+    }
+  },
+
+  calculateGasPriceInEther(gas: bigint, gasPrice: bigint) {
+    const totalGasCostInWei = gasPrice * gas
+    const totalGasCostInEther = Number(totalGasCostInWei) / 1e18
+
+    return totalGasCostInEther
+  },
+
+  calculateGasPriceInUSD(gas: bigint, gasPrice: bigint) {
+    try {
+      const totalGasCostInEther = this.calculateGasPriceInEther(gas, gasPrice)
+      const networkPriceNumber = NumberUtil.bigNumber(state.networkPrice)
+      const totalCostInUSD = networkPriceNumber.multipliedBy(totalGasCostInEther)
+
+      return totalCostInUSD.toNumber()
+    } catch (error) {
+      return 0
+    }
+  },
+
+  calculatePriceImpact(_toTokenAmount: string, _gasPriceInUSD = 0) {
+    const toTokenAmount = NumberUtil.bigNumber(_toTokenAmount)
+
+    const totalSourceCostUSD = NumberUtil.bigNumber(state.sourceTokenAmount).multipliedBy(
+      state.sourceTokenPriceInUSD
+    )
+    const adjustedTotalSourceCostUSD = totalSourceCostUSD.plus(_gasPriceInUSD)
+    const effectivePricePerTargetToken = adjustedTotalSourceCostUSD.dividedBy(
+      toTokenAmount.toFixed(4)
+    )
+
+    const priceImpact = effectivePricePerTargetToken
+      .minus(state.toTokenPriceInUSD)
+      .dividedBy(state.toTokenPriceInUSD)
+      .multipliedBy(100)
+
+    return priceImpact.toNumber()
+  },
+
+  calculateMaxSlippage() {
+    const slippageToleranceDecimal = NumberUtil.bigNumber(state.slippage).dividedBy(100)
+    const maxSlippageAmount = NumberUtil.multiply(state.sourceTokenAmount, slippageToleranceDecimal)
+
+    return maxSlippageAmount.toNumber()
+  },
+
+  async convertTokens() {
+    const { sourceTokenAddress, toTokenAddress } = this.getParams()
+
+    if (!sourceTokenAddress || !toTokenAddress) {
+      return
+    }
+
+    await this.makeChecks()
+  },
+
+  async makeChecks() {
+    const { toTokenDecimals, toTokenAddress } = this.getParams()
+
+    if (!toTokenDecimals || !toTokenAddress) {
+      return
+    }
+
+    state.loading = true
+    const transaction = await this.getTransaction()
+    this.setTransactionDetails(transaction)
+    state.loading = false
+  },
+
+  async getTransaction() {
+    const { fromAddress, sourceTokenAddress, sourceTokenAmount, sourceTokenDecimals } =
+      this.getParams()
+
+    if (
```

Comment:
We should handle this state in UI. User shouldn't be able to run this function

Reply:
I mean if they somehow press button but some of these values are empty, the function returns null which mean do nothing, and if the user don't get any feedback in UI, this wouldn't be good.